### ISSUE TITLE
Local variables not shown if an asynchronous gap in the call stack

### ISSUE THREAD

**Issue by kinex** [Dart-Code/Dart-Code#1507]
Local variables are not shown in debugger if you select in Call Stack -view any line below the topmost `<asynchronous gap>` line. Sample code:

```dart
void main() async {
  final i = 10;
  await func1(i);
}

Future<int> func1(int value) async {
  await Future.delayed(Duration(seconds: 1));
  value = value + 10;
  final value2 = await func2(100);
  final result = value + value2;
  return result;
}

Future<int> func2(int value) async {
  await Future.delayed(Duration(seconds: 1));
  final result = value * 2;
  return result;
}
```
Steps to reproduce:
1. Add a breakpoint to `func2` and start debugging
2. Wait until the breakpoint is hit
3. Select `func1` or `main` in the Call Stack -view and try to see variables
4. Result: local variables are not shown in Variables-view and no tooltips are shown for parameters or other local variables.

Problem disappears if you remove all await/async stuff from the sample code.

**Comment by kinex** [comment]
Ok, thanks! Let me know if you need more info.

**Comment by DanTup** [comment]
I can repro this - and strangely I can inspect the variables if I select the `<async gap>` frame. I think the indexes might be out of sync or something. I'll do some digging!

**Comment by DanTup** [comment]
This turns out to be complicated... The call stack we're showing here is somewhat fake (the `asyncCausalStack`). It's great for display, because it looks more like your code. Unfortunately because the frames aren't real, we don't have access to the locals, and we can't evaluate expressions in them - we're actually showing locals/evaluation results from the *real* stack frames, which you can't see.

I don't think we can support what we'd like here, so we might need to see if we can make a way to toggle between the call stack views, and let you see locals/expressions when using the "real" one but still have the async list for reference.

**Comment by DanTup** [comment]
So, it turns out we can't support evaluation/locals for frames past the first async boundary. I'm told the frames are fabricated and the locals inaccessible (it would be significant VM work to make this work).

I think to tidy this up, the best we can do is disable the expression evaluations and locals once you've gone past the top-most async frame to avoid confusion.

**Comment by kinex** [comment]
I tested this also in Android Studio + Flutter plugin and it doesn't work there either. But for comparison, similar feature works in VS/C# debugger.

I guess one workaround is to set breakpoints before each async boundary when you really need to see all parameters and local variables. This is of course possible only if you can reproduce the exception you are debugging. 

It would be perfect to have this kind of debugging feature, but I guess we can live without it also.

Thank you for investigating this!

**Comment by DanTup** [comment]
Yeah, I knew it worked in C# and I really wished it could work here. Unfortunately it sounds like a limitation of how the VM works (I don't think it was originally built with this in mind). Fixing that is a feature request for https://github.com/dart-lang/sdk, though the impression I get is that it's probably too big a change to really be worthwhile.

I'll keep this issue open for tidying up and making sure we don't evaluate in the wrong frame.

**Comment by DanTup** [comment]
Closing this as a dupe of the SDK issue https://github.com/dart-lang/sdk/issues/51763. It might be that there's some work to do in the debug adapter if the VM gets updated for this, but we can re-open this (or file a DAP issue in the SDK) if that happens.
